### ISSUE TITLE
Add explicit resolution capabilities to bsp

### ISSUE THREAD

**Issue by jvican** [scalacenter/bsp#14]
`buildTarget/dependencySources` is one of the most important requests in BSP. `dependencySources` allows language servers to manage dependencies' indices and provide language features like syntax highlighting, code navigation and autocompletions. These features are supported for both the projects of a build and the library dependencies (when you go to the definition of a method in a library, you want to be able to go to definitions of its body).
  
To achieve this full experience, IDEs extract *all source/doc jars* of the build via their custom integrations (think sbt, maven or gradle integrations in IntelliJ) and index the source code. This is a **slow** process that makes the "import project" experience poor. When developers clone projects and import them, the process can take easily more than 10 minutes because the build tool needs to download all jars (whose size can be in the order of GB for a medium-sized Scala project).
  
All Scala developers that want to contribute to a library for the first time need to go through this expensive process. I've noticed that in my case I've tried hard to avoid cloning projects locally and preferred making changes via the github interface to avoid the wait, but ultimately made the contributing experience even slower without the help of a local IDE to compile and browse.
  
## A better solution
  
The problem here is that the download of jars (binary, source and docs) happens upfront when the user clicks "import project" in the IDE. In most of the cases, the download of source and doc jars are not necessary because users will most likely never browse all dependencies of a project!
  
I think we could address this problem by making the source and doc jar resolution lazy and letting the language server handle it. `dependencySources` would be free not to return any source jars, and IntelliJ, metals or Dotty IDE would ask the build tool to resolve the jar of a dependency when the user opens up a file that hasn't been indexed.
  
This approach will save disk space in people's computers and save time when doing the full import project in the IDE. 
  
This enhancement would require the addition of a `ResolutionProvider` and a `buildTarget/resolve` endpoint (similar to `buildTarget/dependencySources). This endpoint would take a list of URIs (binary jars) and return a list of (resolved) source jars. The idea is that the build tool will only resolve the source jars for the requested binary jars. We may want to make some parts of this language-specific.
  
/cc @jastice

**Comment by olafurpg** [comment]
I am not convinced this is a good idea. Sources are used for more than goto definition, you need to sources to index docstrings and I think (but not 100% sure) IntelliJ scala uses sources to find parameter default values 

![screen shot 2018-06-11 at 12 24 29](https://user-images.githubusercontent.com/1408093/41226649-6935d1c0-6d72-11e8-96dd-527bb2454813.png)

Additionally, users expect goto definition to respond instantly so adding a several second delay to that experience would be a shame.

However, I think BSP already improves on the state of art by separate dependency sources into a custom endpoint. This allows IDEs to get critical information first (unmanagedSources, dependencyClasspath, ...) and then later fetch sources.

**Comment by jvican** [comment]
> I am not convinced this is a good idea. Sources are used for more than goto definition, you also need sources to index docstrings and I think (but not 100% sure) IntelliJ scala uses sources to find parameter default values
  
That's implementation specific, I don't think it makes a good case against this feature. For example, parameter default values could be obtained from scala signatures in bytecode.
  
> Additionally, users expect goto definition to respond instantly so adding a several second delay to that experience would be a shame.
 
Browsing dependencies code is not something that is done often, and so having to resolve all jars and have a slow import project that downloads the whole universe just to satisfy a feature it might not even be used is a pity.
  
I would prefer we optimize the IDE experience for the features that are used most, like a JIT only optimizes the code that is most executed.
  
> Additionally, users expect goto definition to respond instantly so adding a several second delay to that experience would be a shame.
  
There are trade-offs and this is a one-time cost. The UI could show a nice progress bar saying it's resolving the jar. I think this is better than waiting minutes to import the project.
  
Another point in favor of this feature is that IntelliJ can use it to replace the current "Attach Sources" popup window that happens when you browse a dependency that doesn't have sources. Instead of showing that and asking the user to manually select the source jar for that dependency, IntelliJ can ask the build tool about it and the user just needs to wait and not think.

**Comment by olafurpg** [comment]
I see what you mean. I agree this might be a good idea, and one example of how client/server may be better than eagerly persisting all information to JSON/XML on disk. I would let it wait for a future version of BSP (maybe v2), I suspect it requires custom work on both sides (IDEs + build tools) to make this feature work well.

**Comment by jastice** [comment]
Agree, it would take some extra work on our side as well to support on-demand sources, considering they are also used for documentation. I do think a second-or-so delay is okay for intial resolving of a single source. We would have to see how it works out for large source artifacts.

**Comment by retronym** [comment]
@jastice With the existing protocol, would IntelliJ be able to defer the `dependencySources` query to a second pass, i.e. import the project without them, letting the user start editing, and then ask the BSP server for the dependency source JAR download?

**Comment by jastice** [comment]
@retronym technically yes, there even is an existing mechanism for first pass/second pass import, but we effectively skip the first pass so that the user can start editing even if the server fails completely. Adding a third pass to this might be more complicated though, and would be more resource-consuming. Partial updates of imports and attaching sources later is also possible, but makes things more complicated, too.
